node-wasm/src/node.rs

```diff
@@ -38,141 +37,246 @@ pub struct WasmNodeConfig {
     pub bootnodes: Vec<String>,
 }
 
-#[wasm_bindgen(js_class = Node)]
-impl WasmNode {
-    /// Create a new Lumina node.
-    #[wasm_bindgen(constructor)]
-    pub async fn new(config: WasmNodeConfig) -> Result<WasmNode> {
-        let config = config.into_node_config().await?;
+/// `NodeDriver` represents lumina node running in a dedicated Worker/SharedWorker.
+/// It's responsible for sending commands and receiving responses from the node.
+#[wasm_bindgen(js_name = NodeClient)]
+struct NodeDriver {
+    client: WorkerClient,
+}
 
-        if let Ok(store_height) = config.store.head_height().await {
-            info!("Initialised store with head height: {store_height}");
+/// Type of worker to run lumina in. Allows overriding automatically detected worker kind
+/// (which should usually be appropriate).
+#[wasm_bindgen]
+pub enum NodeWorkerKind {
+    /// Run in [`SharedWorker`]
+    ///
+    /// [`SharedWorker`]: https://developer.mozilla.org/en-US/docs/Web/API/SharedWorker
+    Shared,
+    /// Run in [`Worker`]
+    ///
+    /// [`Worker`]: https://developer.mozilla.org/en-US/docs/Web/API/Worker
+    Dedicated,
+}
+
+#[wasm_bindgen(js_class = NodeClient)]
+impl NodeDriver {
+    /// Create a new connection to a Lumina node running in a Shared Worker.
+    /// Note that single Shared Worker can be accessed from multiple tabs, so Lumina may
+    /// already have been started. Otherwise it needs to be started with [`NodeDriver::start`].
+    #[wasm_bindgen(constructor)]
+    pub async fn new(worker_type: Option<NodeWorkerKind>) -> Result<NodeDriver> {
+        let url = worker_script_url();
+        let mut opts = WorkerOptions::new();
+        opts.type_(WorkerType::Module);
+        opts.name(LUMINA_WORKER_NAME);
+
+        let default_worker_type = if is_chrome() {
+            NodeWorkerKind::Dedicated
         } else {
-            info!("Initialized new empty store");
-        }
+            NodeWorkerKind::Shared
+        };
+
+        let client = match worker_type.unwrap_or(default_worker_type) {
+            NodeWorkerKind::Shared => {
+                info!("Starting SharedWorker");
+                let worker = SharedWorker::new_with_worker_options(&url, &opts)
+                    .to_error("could not create SharedWorker")?;
+                WorkerClient::from(worker)
+            }
+            NodeWorkerKind::Dedicated => {
+                info!("Starting Worker");
+                let worker =
+                    Worker::new_with_options(&url, &opts).to_error("could not create Worker")?;
+                WorkerClient::from(worker)
+            }
+        };
+
+        Ok(Self { client })
+    }
+
+    /// Check whether Lumina is currently running
+    pub async fn is_running(&self) -> Result<bool> {
+        let command = NodeCommand::IsRunning;
+        let response = self.client.exec(command).await?;
```

Comment:
shouldn't errors in this function mean the node is not running?

Reply:
That's bordering on philosophical question :P 
I wanted to distinguish situation of "Worker may be running, but definitely isn't responding" and "We can communicate with worker, but it's not running", since they have different recovery procedures - `start`ing the node and trying to re-create a worker.